web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php

```diff
@@ -0,0 +1,269 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\joinup_federation\Plugin\pipeline\Step;
+
+use Drupal\Component\Utility\NestedArray;
+use Drupal\Core\Entity\EntityInterface;
+
+/**
+ * Utility trait to handle information of the federation incoming entities.
+ */
+trait IncomingEntitiesDataHelperTrait {
+
+  /**
+   * A dependency tree for each incoming solution.
+   *
+   * An associative array indexed by the solution ids that contains the
+   * 'dependencies' and the 'category' values. The 'dependencies' is an
+   * associative array itself, of entity ids indexed by their bundle that are
+   * related to the solution. The 'category' is a string representing the status
+   * of the solution. Possible values are 'not_federated', 'federated',
+   * 'invalid_solution', 'federated_unchanged' and 'blacklisted'.
+   *
+   * @var array
+   */
+  protected $solutionData = NULL;
+
+  /**
+   * An associative array of hashes indexed by entity id.
+   *
+   * @var array
+   */
+  protected $entityHashes = NULL;
+
+  /**
+   * Loads the solution data from the persistent state.
+   */
+  protected function ensureEntityDataLoaded(): void {
```

Comment:
It will cause errors only if you do not run it before any processing. The difference is that it will have to be explicitly called every time a step needs any of the versions. Calling in the methods will ensure data are loaded. I can remove it if you want and explicitly call it in the two steps we are calling for it.

Reply:
In general, that seemed a good idea when I started using it. Having a few methods only that were doing the major calculations like setting and retrieving basic properties. However, as I was working on the ticket, it turns out I needed too many methods to have this required in each one of them. Plus I ended up having methods required only by one step, living in the trait due to this method. This method is now renamed and called once in every step needed.